src/udev/udevd.c

```diff
@@ -545,25 +555,23 @@ static void worker_spawn(Manager *manager, struct event *event) {
 }
 
 static void event_run(Manager *manager, struct event *event) {
-        struct worker *worker;
-        Iterator i;
-
         assert(manager);
         assert(event);
-
-        HASHMAP_FOREACH(worker, manager->workers, i) {
+retry:
+        if (!LIST_IS_EMPTY(manager->idle_workers)) {
+                struct worker *worker = manager->idle_workers;
                 ssize_t count;
 
-                if (worker->state != WORKER_IDLE)
-                        continue;
+                assert(worker->state == WORKER_IDLE);
```

Comment:
Can this be handled more gracefully?

What's the rationale for this change, does it make udev measurably faster? My fear is that this double bookkeeping makes things more brittle, in particular forgetting to update the list in all possible places. Such assertions would then make the main udevd crash, and bad things happen.